File: R/main_collector.R
# Script principal de collecte des données.
# Sources: FAOSTAT, Open-Meteo, Agridata, INM meteo.tn, Sol synthétique.
#
# Usage: Rscript R/main_collector.R [dossier_sortie]
# Par défaut: agriculture_scraping/data/

script_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) "R")

source(file.path(script_dir, "scraping", "faostat_api.R"))
source(file.path(script_dir, "scraping", "openmeteo_api.R"))
source(file.path(script_dir, "scraping", "agridata_scraper.R"))
source(file.path(script_dir, "scraping", "meteo_scraper.R"))
source(file.path(script_dir, "scraping", "soil_data_generator.R"))
source(file.path(script_dir, "preprocessing", "data_merger.R"))


run_collection <- function(output_dir) {
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  summary_out <- list()

  # 1) FAOSTAT
  cat("--- FAOSTAT (rendement, production, surface) ---\n")
  tryCatch({
    df_fao <- download_faostat_crops_tunisia(output_dir)
    summary_out$faostat <- nrow(df_fao)
    cat(sprintf("  Lignes: %d\n", nrow(df_fao)))
  }, error = function(e) {
    summary_out$faostat <<- 0
    cat(sprintf("  Erreur: %s\n", conditionMessage(e)))
  })

  # 2) Open-Meteo
  cat("--- Open-Meteo (température, pluie, humidité) ---\n")
  tryCatch({
    df_om <- save_openmeteo_tunisia(output_dir)
    summary_out$openmeteo <- nrow(df_om)
    cat(sprintf("  Lignes: %d\n", nrow(df_om)))
  }, error = function(e) {
    summary_out$openmeteo <<- 0
    cat(sprintf("  Erreur: %s\n", conditionMessage(e)))
  })

  # 3) Agridata
  cat("--- Agridata (pluviométrie, production) ---\n")
  tryCatch({
    df_pluv <- scrape_agridata_pluviometry(output_dir)
    df_prod <- scrape_agridata_production(output_dir)
    summary_out$agridata_pluviometry <- nrow(df_pluv)
    summary_out$agridata_production  <- nrow(df_prod)
    cat(sprintf("  Pluviométrie: %d, Production: %d\n", nrow(df_pluv), nrow(df_prod)))
  }, error = function(e) {
    summary_out$agridata_pluviometry <<- 0
    summary_out$agridata_production  <<- 0
    cat(sprintf("  Erreur: %s\n", conditionMessage(e)))
  })

  # 4) INM meteo.tn
  cat("--- INM meteo.tn ---\n")
  tryCatch({
    df_inm <- collect_inm_data(output_dir)
    summary_out$meteo_tn <- nrow(df_inm)
    cat(sprintf("  Lignes: %d\n", nrow(df_inm)))
  }, error = function(e) {
    summary_out$meteo_tn <<- 0
    cat(sprintf("  Erreur: %s\n", conditionMessage(e)))
  })

  # 5) Données sol synthétiques
  cat("--- Données sol (pH, N, P, K) ---\n")
  tryCatch({
    df_soil <- save_soil_data(output_dir)
    summary_out$soil <- nrow(df_soil)
    cat(sprintf("  Lignes: %d\n", nrow(df_soil)))
  }, error = function(e) {
    summary_out$soil <<- 0
    cat(sprintf("  Erreur: %s\n", conditionMessage(e)))
  })

  # 6) Fusion pour ML
  cat("--- Fusion des données ---\n")
  tryCatch({
    merged_path <- merge_all_to_ml_dataset(output_dir)
    summary_out$merged_ml <- merged_path
    cat(sprintf("  Fichier fusionné: %s\n", merged_path))
  }, error = function(e) {
    summary_out$merged_ml <<- NULL
    cat(sprintf("  Fusion: %s\n", conditionMessage(e)))
  })

  summary_out
}


if (!interactive()) {
  args <- commandArgs(trailingOnly = TRUE)
  out  <- if (length(args) >= 1) args[1] else file.path(getwd(), "agriculture_scraping", "data")
  cat(sprintf("Sortie: %s\n\n", out))
  run_collection(out)
  cat("\nCollecte terminée.\n")
}

File: agriculture_scraping/scripts/main_collector.py
# -*- coding: utf-8 -*-
"""
Script principal de collecte des données pour le projet
"Prédiction du rendement agricole à partir des conditions climatiques et du sol en Tunisie".

Sources:
  - FAOSTAT: rendement, production, surface cultivée
  - Open-Meteo: température, précipitations, humidité
  - Agridata.tn: pluviométrie, production (si disponible)
  - INM meteo.tn: données climatiques (complément)
  - Données sol (générées): pH, N, P, K

Usage:
  python scripts/main_collector.py [dossier_sortie]
  Par défaut: data/
"""

import sys
from pathlib import Path

# Ajouter la racine du projet au path
ROOT = Path(__file__).resolve().parent.parent
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from scripts.scraping import agridata_scraper, faostat_api, meteo_scraper, openmeteo_api, soil_data_generator


def run_collection(output_dir: str | Path) -> dict:
    """
    Lance toute la collecte et enregistre les CSV dans output_dir.
    Retourne un résumé par source.
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    summary = {}

    # 1) FAOSTAT - rendement, production, surface
    print("--- FAOSTAT (rendement, production, surface) ---")
    try:
        df_fao = faostat_api.download_faostat_crops_tunisia(output_dir)
        summary["faostat"] = len(df_fao)
        print(f"  Lignes: {len(df_fao)}")
    except Exception as e:
        summary["faostat"] = 0
        print(f"  Erreur: {e}")

    # 2) Open-Meteo - climat
    print("--- Open-Meteo (température, pluie, humidité) ---")
    try:
        df_om = openmeteo_api.save_openmeteo_tunisia(output_dir)
        summary["openmeteo"] = len(df_om)
        print(f"  Lignes: {len(df_om)}")
    except Exception as e:
        summary["openmeteo"] = 0
        print(f"  Erreur: {e}")

    # 3) Agridata - pluviométrie / production
    print("--- Agridata (pluviométrie, production) ---")
    try:
        df_pluv = agridata_scraper.scrape_agridata_pluviometry(output_dir)
        df_prod = agridata_scraper.scrape_agridata_production(output_dir)
        summary["agridata_pluviometry"] = len(df_pluv)
        summary["agridata_production"] = len(df_prod)
        print(f"  Pluviométrie: {len(df_pluv)}, Production: {len(df_prod)}")
    except Exception as e:
        summary["agridata_pluviometry"] = summary["agridata_production"] = 0
        print(f"  Erreur: {e}")

    # 4) INM meteo.tn
    print("--- INM meteo.tn ---")
    try:
        df_inm = meteo_scraper.collect_inm_data(output_dir)
        summary["meteo_tn"] = len(df_inm)
        print(f"  Lignes: {len(df_inm)}")
    except Exception as e:
        summary["meteo_tn"] = 0
        print(f"  Erreur: {e}")

    # 5) Données sol (générées)
    print("--- Données sol (pH, N, P, K) ---")
    try:
        df_soil = soil_data_generator.save_soil_data(output_dir)
        summary["soil"] = len(df_soil)
        print(f"  Lignes: {len(df_soil)}")
    except Exception as e:
        summary["soil"] = 0
        print(f"  Erreur: {e}")

    # 6) Fusion pour ML (si preprocessing disponible)
    print("--- Fusion des données ---")
    try:
        from scripts.preprocessing.data_merger import merge_all_to_ml_dataset
        merged_path = merge_all_to_ml_dataset(output_dir)
        summary["merged_ml"] = str(merged_path)
        print(f"  Fichier fusionné: {merged_path}")
    except Exception as e:
        summary["merged_ml"] = None
        print(f"  Fusion: {e}")

    return summary


if __name__ == "__main__":
    out = sys.argv[1] if len(sys.argv) > 1 else str(ROOT / "data")
    print(f"Sortie: {out}\n")
    run_collection(out)
    print("\nCollecte terminée.")
